slatedb/src/merge_operator.rs

```diff
@@ -138,94 +184,149 @@ impl<T: KeyValueIterator> MergeOperatorIterator<T> {
+            ValueDeletable::Value(bytes) => Some(bytes.clone()),
+            _ => None,
+        };
+        let mut found_base_value = !matches!(first_entry.value, ValueDeletable::Merge(_));
+
+        let mut batch_results: Vec<Bytes> = Vec::new();
+
+        // Stream and process entries in batches of MERGE_BATCH_SIZE
+        let mut batch = Vec::with_capacity(MERGE_BATCH_SIZE);
+
+        let mut done = self.process_merge_entry(
+            first_entry,
+            &mut batch,
+            &mut base_value,
+            &mut found_base_value,
+            &mut max_create_ts,
+            &mut min_expire_ts,
+            &mut seq,
+        );
+
+        while !done {
+            while batch.len() < MERGE_BATCH_SIZE && !done {
+                let Some(next_entry) = self.delegate.next_entry().await? else {
+                    done = true;
                     break;
-                }
-                None => {
-                    // End of iterator
+                };
+
+                // Check if entry matches current key and expire_ts
+                if !self.is_matching_entry(&next_entry, &key, first_expire_ts) {
+                    self.buffered_entry = Some(next_entry);
+                    done = true;
                     break;
                 }
+
+                // Validate sequence number ordering (descending)
+                let last_seq = batch.last().map(|(s, _)| *s).unwrap_or(seq);
+                self.validate_sequence_order(last_seq, next_entry.seq)?;
+
+                done = self.process_merge_entry(
+                    next_entry,
+                    &mut batch,
+                    &mut base_value,
+                    &mut found_base_value,
+                    &mut max_create_ts,
+                    &mut min_expire_ts,
+                    &mut seq,
+                );
             }
-        }
 
-        // Reverse entries so we merge from oldest to newest
-        entries.reverse();
+            if !batch.is_empty() {
+                batch.reverse();
 
-        // don't call the merge operator if the base value is a value
-        // and instead just set it as the initial merge value
-        let mut merged_value: Option<Bytes> =
-            if let ValueDeletable::Value(bytes) = &entries[0].value {
-                Some(bytes.clone())
-            } else {
-                None
-            };
-
-        let mut max_create_ts = entries[0].create_ts;
-        let mut min_expire_ts = entries[0].expire_ts;
-        let mut seq = entries[0].seq;
-
-        // a base value can be either a tombstone or a value, in either
-        // case we should not apply the merge operator to it (in the former
-        // case we just apply merges and in the latter we have already set
-        // the merged value to the base value)
-        if found_base_value {
-            entries.remove(0);
-        }
+                let operands: Vec<Bytes> = batch.iter().map(|(_, v)| v.clone()).collect();
```

Comment:
not a regression so we can merge without this, but it's a bummer we have to clone all the values 🤔 I wonder if there's a way we can get around that